conn.go

```diff
@@ -347,6 +380,93 @@ func (conn *Conn) inWorker() {
 	}
 }
 
+// sigWorker gets signals from signal queue and handles them
+// appropriately.
+func (conn *Conn) sigWorker() {
+	for {
+		signal := conn.nextSignal()
+		if signal != nil {
+			closing := conn.handleSignal(signal)
+			if closing {
+				return
+			}
+		} else {
+			select {
+			case <-conn.signalQueueWakeUp:
+			case <-conn.shutdownSigWorker:
+				return
+			}
+		}
+	}
+}
+
+// nextSignal gets next signal from queue or nil if there was no
+// signal in queue.
+func (conn *Conn) nextSignal() *Signal {
+	conn.signalQueueLck.Lock()
+	defer conn.signalQueueLck.Unlock()
+	for len(conn.signalQueue) > 0 {
+		signal := conn.signalQueue[0]
+		conn.signalQueue = conn.signalQueue[1:]
+		if signal != nil {
+			return signal
+		}
+	}
+
+	conn.signalQueueConsumerAsleep = true
+	return nil
+}
+
+// processPendingSignalChannels either adds new pending channels to
+// signal channels list or removes them if they already existed there.
+//
+// It is assumed that conn.signalsLck is locked.
+func (conn *Conn) processPendingSignalChannels() {
+	signalsMap := make(map[chan<- *Signal]bool, len(conn.signals))
+	for _, ch := range conn.signals {
+		signalsMap[ch] = false
+	}
+	conn.pendingSignalChannelsLck.Lock()
+	for _, ch := range conn.pendingSignalChannels {
+		if _, ok := signalsMap[ch]; ok {
+			delete(signalsMap, ch)
+		} else {
+			signalsMap[ch] = false
+		}
+	}
+	conn.pendingSignalChannels = nil
+	conn.pendingSignalChannelsLck.Unlock()
+
+	conn.signals = nil
+	for ch, _ := range signalsMap {
+		conn.signals = append(conn.signals, ch)
```

Comment:
My original comment was getting at the issue of not knowing how channels act when used in a `range`. To illustrate my example, consider [0] compared to [1]. In example [0], the resulting slice ends up with three pointers all pointing to the same memory address. In example [1] (which simulates your code), the resulting object looks like we expect. My suggestion that we might "need to copy `ch`" is illustrated by the commented out line in example [0] where `v` is copied before a pointer is created and appended to the slice.

So, it took all this to say LGTM :)

[0] http://play.golang.org/p/iV0lurQvgq
[1] http://play.golang.org/p/0jlrm7ZMUk